### ISSUE TITLE
Hierarchical grouping of data

### ISSUE THREAD

**Issue by Void9711** [marcusolsson/grafana-treemap-panel#5]
Hello, I've imported little amout of stats from elasticsearch into grafana and chose this plugin to draw a treemap panel.

Grafana version: **7.0.6**
Treemap plugin version: **0.2.0**
Display: **Squarify**
Dimensions:
- Text: **asset**
- Size: **size**
- Color: **_index**

It performs as below:
![image](https://user-images.githubusercontent.com/48246306/90587815-d8065500-e20c-11ea-8358-d4f0a18f16a7.png)

### Expected result: The 'Engine/' tier will include its subdirectories instead of taking a bulk standing alone. ###
Just like the visualization of the treemap lib in [echarts](https://echarts.apache.org/examples/en/#chart-type-treemap).
Is there anything wrong with my settings or it's just an uncompleted function of treemap plugin?

My stats in elasticsearch:
```json
{
  "_index" : "game-resource-pirates-1",
  "_type" : "_doc",
  "_id" : "1",
  "_version" : 8,
  "_seq_no" : 17,
  "_primary_term" : 3,
  "found" : true,
  "_source" : {
    "asset" : "Engine/Content/xxx/1.uasset",
    "offset" : 0,
    "size" : 413,
    "sha1" : "154E806131D7C77245C00088593979C510407E7C",
    "@timestamp" : "2020-08-12T10:06:56Z"
  }
}
```

```json
{
  "_index" : "game-resource-pirates-1",
  "_type" : "_doc",
  "_id" : "2",
  "_version" : 4,
  "_seq_no" : 18,
  "_primary_term" : 3,
  "found" : true,
  "_source" : {
    "asset" : "Engine/Content/yyy/2.uasset",
    "offset" : 486,
    "size" : 435,
    "sha1" : "DB784BF38D47A6EFDA1E131064D45065584EA8A5",
    "@timestamp" : "2020-08-12T10:06:59Z"
  }
}
```

```json
{
  "_index" : "game-resource-pirates-1",
  "_type" : "_doc",
  "_id" : "3",
  "_version" : 4,
  "_seq_no" : 19,
  "_primary_term" : 3,
  "found" : true,
  "_source" : {
    "asset" : "Engine/Content/zzz/3.uasset",
    "offset" : 994,
    "size" : 443,
    "sha1" : "F4BD3E99142AF63A07A871562070F09130B93CC3",
    "@timestamp" : "2020-08-12T10:07:04Z"
  }
}
```

```json
{
  "_index" : "game-resource-pirates-1",
  "_type" : "_doc",
  "_id" : "5",
  "_version" : 3,
  "_seq_no" : 20,
  "_primary_term" : 3,
  "found" : true,
  "_source" : {
    "@timestamp" : "2020-08-12T10:07:08Z",
    "asset" : "Engine/Content/xyz/5.uasset",
    "offset" : 71302,
    "size" : 1048,
    "sha1" : "253084F0D6206E5DCE7540E06A156BB740187A13"
  }
}
```

```json
{
  "_index" : "game-resource-pirates-1",
  "_type" : "_doc",
  "_id" : "6",
  "_version" : 2,
  "_seq_no" : 21,
  "_primary_term" : 3,
  "found" : true,
  "_source" : {
    "@timestamp" : "2020-08-12T10:07:08Z",
    "asset" : "Engine/",
    "offset" : 87054,
    "size" : 2339,
    "sha1" : "BEC945B4758C4D2E57CDBEF6FF12D205A4014A1E"
  }
}
```

My query in grafana:
![image](https://user-images.githubusercontent.com/48246306/90599034-b9ad5300-e226-11ea-8273-370b7597029d.png)

**Comment by marcusolsson** [comment]
Thank you for taking the time to report this! This is indeed a limitation of the current treemap: It only supports one level. The underlying function from the D3 library supports this so it should be straightforward to add.

I assume this would require adding a `parent` field to the query that specifies which box to add it under.

**Comment by Void9711** [comment]
Got it. Thanks!
How long could this to be added into treemap plugin and release?
And is it possible to attach an optional hyperlink on every data bulk to provide a quick jump?
: )

**Comment by marcusolsson** [comment]
I don't know when I can look at this at the moment unfortunately. If anyone wants to contribute this, I'd happily help out!

**Comment by jermudgeon** [comment]
I've been using https://grafana.com/grafana/plugins/digrich-bubblechart-panel/ for a similar purpose when what I really want is a hierarchical squarified treemap. That plugin uses a fairly simple single concatenated field: "root,hierarchy1,hierarchy2,node" for example, allowing arbitrary depths and nesting. Perhaps that would help? Currently when I try to use Group here I get an error.

**Comment by MaxMcGlinnPoole** [comment]
+1 to this, this would be an extremely helpful addition. We're looking at using this at my work but the lack of support for hierarchical data is limiting.

**Comment by marcusolsson** [comment]
This is definitely something that the plugin should support.

I did a proof-of-concept on this a while back, but unfortunately got stuck and decided to spend time on easier wins for a while. But I'd like to revisit this again at some point. 

One of the thing I believe I couldn't decide on was how you'd actually configure it. For example, would you want to configure a field that would define the parent for each row, or would you rather select fields that you'd like to group by?

**Comment by MaxMcGlinnPoole** [comment]
Yeah a field defining the parent for each row would be the best for my use case. Kinda like the sunburst viz: https://github.com/Dikshita25/sunburst-grafana-plugin

Its a very important feature for our use case. We’re using another plugin in the mean time to visualize our data but we like this plugin a lot more. If you have time to get to this it would be greatly appreciated.

Thanks so much!

**Comment by jermudgeon** [comment]
Re: how to configure parent/child relationships, I was picturing something like 
https://grafana.com/grafana/plugins/digrich-bubblechart-panel/
Bubble Chart uses a single field with a user-defined separator for (presumably) arbitrary depths of aggregation.
"Level1,Level2",metric,data
"Level 1,Level2,Level3",metric,data
I find that this makes it easy to pass arbitrary depths without having to have a complex graph-assembling data source. I can do the field concatenation in the query (BigQuery, in my case).

**Comment by marcusolsson** [comment]
Made some progress on this, but there are still work to be done. Is this what you'd expect from the data in the issue description?

<img width="679" alt="Screenshot 2021-05-12 at 22 53 59@2x" src="https://user-images.githubusercontent.com/8396880/118042911-f69d1500-b374-11eb-8554-564b47dae7d9.png">

One thing I've run into though: In the example in the issue description, `Engine/` has a size even though it's not a leaf node. The way D3.js works is that it sums the value for each node. If the parent has it's own value, it will be added to the sum of all its children, which is probably not what you want in this case.

Those of you who've been using other treemaps, how would you expect this to be handled? Should parent values be ignored instead?

**Comment by MaxMcGlinnPoole** [comment]
I’d expect a way to look at values at different parts of the tree. So I’m focusing on the lowest level of my hierarchy I would want parent values to be ignored. If I’m at the next step up the tree I would expect the parent’s child values to be ignored. I have values for my parent that are almost equal to the sum of the children but aren’t exactly the same.

**Comment by marcusolsson** [comment]
> I’d expect a way to look at values at different parts of the tree.

I think this would be a good feature to have, though I think it should be out of scope for now. I think we can scope this feature request down to a static presentation of hierarchical data, and add a "drill-down" feature as a follow-up.

> I have values for my parent that are almost equal to the sum of the children but aren’t exactly the same.

This is more of the problem I've having right now. In the case where the sum of the children doesn't equal the value of the parent. Should one override the other? We can always make this configurable, but I'd like to find a good default for now.

**Comment by MaxMcGlinnPoole** [comment]
My thought is that if the parent value is different than the sum of its children, then the parent value should override the sum of the children. As you say, you can make that configurable later and before making that configurable users do have a work around. They can just format their data differently so that the parent value is equal to the sum of its child value. In the other approach (overriding parent value for sum of children) there isn’t really a workaround to that if a user wants the parent value to override the sum of its children.

**Comment by marcusolsson** [comment]
Another status update. I decided to go with configuring by specifying a path in the text dimension with a configurable separator. A couple reasons:

- Using paths keeps reuses the text dimension. An empty separator value recreates the current functionality.
- Users don't have to "engineer" the data source query for the visualization. The hierarchy is configured in the panel rather than in the data. 

This solution also continues to work nicely with the Group by dimension. We can discuss adding a Parent dimension in a later pull request, but I think this is a good start. I'll clean this up and hopefully we can release this during the week.

![treemap](https://user-images.githubusercontent.com/8396880/118540225-ad700b00-b750-11eb-86c2-35be196e8544.gif)

**Comment by MaxMcGlinnPoole** [comment]
Great! Thanks for the update

**Comment by marcusolsson** [comment]
Available in 0.9.0
